### PR TITLE
Public Cache interfaces

### PR DESCRIPTION
**Pull request by icewind1991**
Adds public interfaces for the various cache classes and exposes them on the public storage interface.

Also changes how cache entries are returned, instead of using an array it now returns a class to have a bit more structure to the return values.

cc @PVince81 @DeepDiver1975

### CODE REVIEW

File: lib/private/files/cache/scanner.php [hunk 1]
```diff
@@ -111,7 +106,7 @@ public function setUseTransactions($useTransactions) {
 	 * @param string $path
 	 * @return array an array of metadata of the file
```
Comment (PVince81) [inline lib/private/files/cache/scanner.php]: ICacheEntry[] ?

File: lib/private/files/cache/updater.php [hunk 2]
```diff
@@ -146,18 +147,18 @@ public function remove($path) {
 	/**
 	 * Rename a file or folder in the cache and update the size, etag and mtime of the parent folders
 	 *
-	 * @param \OC\Files\Storage\Storage $sourceStorage
+	 * @param \OCP\Files\Storage $sourceStorage
 	 * @param string $source
 	 * @param string $target
 	 */
-	public function renameFromStorage(\OC\Files\Storage\Storage $sourceStorage, $source, $target) {
+	public function renameFromStorage(\OCP\Files\Storage $sourceStorage, $source, $target) {
```
Comment (PVince81) [inline lib/private/files/cache/updater.php]: I hope one day we can have a IStorage interface :wink:

Comment (PVince81) [inline lib/private/files/cache/updater.php]: maybe now is the opportunity to add it ? Just make it extend `\OCP\Files\Storage` and deprecate the old one ?

File: lib/public/files/cache/icache.php [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+ * Metadata cache for a storage
+ *
+ * The cache stores the metadata for all files and folders in a storage and is kept up to date trough the following mechanisms:
+ *
+ * - Scanner: scans the storage and updates the cache where needed
+ * - Watcher: checks for changes made to the filesystem outside of the ownCloud instance and rescans files and folder when a change is detected
+ * - Updater: listens to changes made to the filesystem inside of the ownCloud instance and updates the cache where needed
+ * - ChangePropagator: updates the mtime and etags of parent folders whenever a change to the cache is made to the cache by the updater
+ *
+ * @since 9.0.0
```
Comment (PVince81) [inline lib/public/files/cache/icache.php]: Kudos for the big phpdoc :smile:

File: lib/public/files/cache/iscanner.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCP\Files\Cache;
+
+/**
+ * Scan files form the storage and save to the cache
```
Comment (PVince81) [inline lib/public/files/cache/iscanner.php]: s/form/from/

File: lib/public/files/cache/iscanner.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	 * @param bool $lock set to false to disable getting an additional read lock during scanning
+	 * @return array an array of metadata of the scanned file
+	 * @throws \OC\ServerNotAvailableException
+	 * @throws \OCP\Lock\LockedException
+	 * @since 9.0.0
+	 */
+	public function scanFile($file, $reuseExisting = 0, $parentId = -1, $cacheData = null, $lock = true);
+
+	/**
+	 * scan a folder and all it's children
```
Comment (PVince81) [inline lib/public/files/cache/iscanner.php]: => its

**Comment by PVince81** [comment]
Nice one!

- [ ] TODO: fix the comments @icewind1991 
- [ ] TODO: smash it

File: lib/private/files/cache/scanner.php (hunk 1 above)
Comment (icewind1991) [inline lib/private/files/cache/scanner.php]: nope, result of $storage->getData(), which is an array

File: lib/public/files/cache/icache.php (hunk 3 above)
Comment (labkode) [inline lib/public/files/cache/icache.php]: As this is the interface for all Caches, the update mechanisms outlined above: Scanner, Watcher, Updater and ChangePropagator should have an empty implementation for Caches that don't use them (management of this mechanism is controlled not by the Cache) or is it the Cache responsible to create and manage them? if the latter then the phodoc should be moved to the implementation.

**Comment by labkode** [comment]
@karlitschek @icewind1991 @DeepDiver1975  @moscicki @dragotin @PVince81 

I really like the refactorisation done around the Cache: **GOOD JOB!!!**

This refactorisation makes easier to plug custom caches, the ICache and the ICacheEntry is a big
:+1:  :+1:  :+1:  :+1:  :+1:  :+1:  :+1:  :+1:  :+1:  :+1:  :+1:  :+1: 

With this pull request our Cache implementation that has been implemented commenting blocks of code here and there with custom calls to EOS, now can reside in a clean class that can fit into the ownCloud architecture avoiding patching some parts of the core.

Nevertheless, this do not solve completely the problem of the Storage Interface we have been discussing. I really want to believe IStorage is not the outcome of #20394 :disappointed: 

I also want to recall the following points:

- Why cannot the file id and storage id be a string ? Enforcing the id to be a number limits future extension of custom caches. You cannot use UUIDs or storage generated ids from DropBox, S3 and family.
 On the other hand, converting a string to an integer to match MySQL auto incremental integer IDs is not going to become a performance bottleneck.
I asked for such feature to the client team and the feature has been included in recent versions.


- If the Cache implementation manages its update mechanisms I would not put in the interface methods and constants that are attached to Caches implementations that need server-to-storage sync. For example: SHALLOW,… getStatus,getIncomplete …)

Now a question it deserves some special attention:

If the backend Storage do not implement the locking mechanisms described in the IStorage, what are the consequences from the Web application point of view and from the sync client ? If this is left in the interface I guess is mandatory to implement it in a correct way, if not, it should be removed.

**Comment by PVince81** [comment]
Note: the string id part is the next step to look into.

@icewind1991 can you answer the other questions ?

**Comment by DeepDiver1975** [comment]
> Nevertheless, this do not solve completely the problem of the Storage Interface we have been discussing. I really want to believe IStorage is not the outcome of #20394 :disappointed:

@labkode as already explained in private communication - we are not willing to take the risk of introducing a totally new interface layer. In an ideal (academic?) world one would throw it away and rebuilt it from scratch - but in the real world we have to ensure that all installations out there continue to work properly.

**Comment by moscicki** [comment]
I understand that this is a first step of a series to bring us to a clear final design (#20394).

We tried to understand how to plug in our storage based on this pull request. Current CERNBox is based on the IObjectStore (https://github.com/owncloud/core/blob/master/lib/public/files/objectstore/iobjectstore.php), not on the local one (https://github.com/owncloud/core/blob/master/lib/public/files/storage.php). This pull requests suggest that the local and remote interfaces are fused into a single one. Looking in this pull request IStorage interface here are some questions. 
 -  what is the meaning of opendir() for remote storage?
 -  is implementing locking mandatory? our system does not need locks.
 - our assumption is that we do not need to implement any of IWatcher, IUpdater and IScanner because in ours case the underlying storage assures the consistency of metadata for all access methods

To get some practical experience with this new interface we would like to give it a try to interface it with our storage as early as possible. What should we do in order to get a working system (checkout a branch?). I understand that this is very alpha stage.

CC: @karlitschek

**Comment by DeepDiver1975** [comment]
> What should we do in order to get a working system (checkout a branch?).

Yes - checking out this branch and start to integrate EOS would be the best from my understanding

**Comment by icewind1991** [comment]
> what is the meaning of opendir() for remote storage?

You can create a "fake" directory handle using

```php
IteratorDirectory::wrap(['file1.txt','file2.png', ...);
```

> is implementing locking mandatory? our system does not need locks.

No, you can just do nothing when we call `acquireLock`

> our assumption is that we do not need to implement any of IWatcher, IUpdater and IScanner because in ours case the underlying storage assures the consistency of metadata for all access methods

You'll need to return a *dummy* implementation of those classes (might add one to core, you can check https://github.com/icewind1991/files_nullcache for examples) from the `getUpdater`, `getScanner`, etc.

I'll try and trim down the interfaces down a bit more from things that don't need to be public (will see if I have time for it this week)

**Comment by icewind1991** [comment]
- Removed a few unneeded methods
- Split of locking methods to a separate interface which can be used to denote that a storage requires oc to do the locking

**Comment by DeepDiver1975** [comment]
rebased

**Comment by DeepDiver1975** [comment]
@icewind1991 can you please have a look at the failing unit tests - THX

**Comment by DeepDiver1975** [comment]
@icewind1991 here is the list of failing unit tests:

````
10:37:50 
10:37:50 There were 5 failures:
10:37:50 
10:37:50 1) OCA\DAV\Tests\Unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testUploadOverWriteReadLocked
10:37:50 Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
10:37:50 
10:37:50 2) OCA\DAV\Tests\Unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testUploadOverWriteWriteLocked
10:37:50 Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
10:37:50 
10:37:50 3) OCA\DAV\Tests\Unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testChunkedUploadOutOfOrderReadLocked
10:37:50 Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
10:37:50 
10:37:50 4) OCA\DAV\Tests\Unit\Connector\Sabre\RequestTest\EncryptionUploadTest::testChunkedUploadOutOfOrderWriteLocked
10:37:50 Failed asserting that exception of type "\OCA\DAV\Connector\Sabre\Exception\FileLocked" is thrown.
10:37:50 
10:37:50 5) OCA\Files_sharing\Tests\Locking::testLockAsRecipient
10:37:50 Failed asserting that exception of type "\OCP\Lock\LockedException" is thrown.
10:37:50

**Comment by icewind1991** [comment]
All green, @PVince81 @DeepDiver1975 please review

File: lib/public/files/cache/icache.php [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+	 * Search for files by tag of a given users.
+	 *
+	 * Note that every user can tag files differently.
+	 *
+	 * @param string|int $tag name or tag id
+	 * @param string $userId owner of the tags
+	 * @return ICacheEntry[] file data
+	 * @since 9.0.0
+	 */
+	public function searchByTag($tag, $userId);
```
Comment (PVince81) [inline lib/public/files/cache/icache.php]: I wonder if we should move the "search" ones to a separate interface ?
But then what happens if we need searching and the storage doesn't implement it ? Use a less efficient workaround ?

**Comment by PVince81** [comment]
Code looks good, see my thoughts about searchByTag, probably not a big issue.

Did some quick test, especially with favorites and cross-storage move, still works. :+1:

File: lib/public/files/cache/icache.php (hunk 6 above)
Comment (DeepDiver1975) [inline lib/public/files/cache/icache.php]: valid - maybe follow up pr :question:

Comment (icewind1991) [inline lib/public/files/cache/icache.php]: Not sure about moving it to a seperate interface, if we do we need to be able to provide a sensible default for backends not implementing the interface which is easy for locking (no locking)

**Comment by PVince81** [comment]
@DeepDiver1975 @nickvergessen @rullzer @schiesbn can we merge this ?

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
@PVince81 let's close all related issues - any followup comments/requests should go into new issues

**Comment by DeepDiver1975** [comment]
> You'll need to return a dummy implementation of those classes (might add one to core, you can check https://github.com/icewind1991/files_nullcache for examples) from the getUpdater, getScanner, etc.

@icewind1991 can we extend your nullcache to be a true storage implementation? We need that as example code for people who want to integrate - THX

@karlitschek ^ as requested :wink:

**Comment by karlitschek** [comment]
Thanks.
@icewind It would be awesome to have a rough example implementation. Doesn't have to be perfect. But something that we can point people to as example.

**Comment by PVince81** [comment]
Added to checklist: https://github.com/owncloud/core/issues/20401#issue-115881984

**Comment by karlitschek** [comment]
Thanks @PVince81

**Pull request merged**